cvmfs/catalog_mgr_rw.cc

```diff
@@ -885,6 +885,77 @@ void WritableCatalogManager::RemoveNestedCatalog(const string &mountpoint,
 }
 
 
+/**
+ * Swap in a new nested catalog
+ *
+ * @param mountpoint - the path of the nested catalog to be removed
+ * @param new_hash - the hash of the new nested catalog
+ * @param new_size - the size of the new nested catalog
+ */
+void WritableCatalogManager::SwapNestedCatalog(const string &mountpoint,
+                                               const shash::Any &new_hash,
+                                               const uint64_t new_size) {
+  const string nested_root_path = MakeRelativePath(mountpoint);
+  const string parent_path = GetParentPath(nested_root_path);
+  const PathString nested_root_ps = PathString(nested_root_path);
+
+  SyncLock();
+
+  // Find the immediate parent catalog
+  WritableCatalog *parent = NULL;
+  if (!FindCatalog(parent_path, &parent)) {
+    PANIC(kLogStderr,
+          "failed to find parent catalog '%s': "
+          "mountpoint was not found in current catalog structure",
+          parent_path.c_str());
+  }
+
+  // Get old catalog hash
+  shash::Any old_hash;
+  uint64_t old_size;
+  bool old_found = parent->FindNested(nested_root_ps, &old_hash, &old_size);
+  if (!old_found) {
+    PANIC(kLogStderr,
+          "failed to swap nested catalog '%s': not found in parent",
+          nested_root_path.c_str());
+  }
+
+  // Check that old catalog was not already attached
+  if (parent->FindChild(nested_root_ps)) {
+    PANIC(kLogStderr,
+          "failed to swap nested catalog '%s': already attached",
+          nested_root_path.c_str());
+  }
+
+  // Load freely attached old catalog
+  UniquePtr<Catalog> old_catalog(LoadFreeCatalog(PathString("", 0), old_hash));
```

Comment:
The `mountpoint` passed to `LoadFreeCatalog` should be the nested path, I think.

Reply:
Thanks, have updated.